### DISCUSSION TITLE
Lesson 4 Doubt in Fund.sol

### DISCUSSION THREAD

**Discussion by rakesh0x7** [Cyfrin/foundry-full-course-cu#208] [Q&A] [answered]
In `Fund.sol` contract we have withdraw function which is used to withdraw funds from the contract but after withdrawing the funds we are clearing the funders array using ` funders = new address[](0);`

**Doubt** : Suppose there are two people in the funders array and funder 1 wants to withdraw the funds using withdraw function 

```   function withdraw() public onlyOwner {
        for (uint256 funderIndex=0; funderIndex < funders.length; funderIndex++){
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
       
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    } 
```
so looking at the function the `for loop` makes funder 1 funds to zero and after that `funders = new address[](0);` this code will make entire array reset which includes funder 2 data to be reseted.

is this logic wrong or I'm thinking in wrong perspective or it works only for one user?

**Comment by DevSwayam** [comment]
So the Above Code is part of the ```withdraw()``` function of smart contract. 
The main motive of function is the owner should be able to withdraw all the funds from the FundMe contract. 
Hence we are using for loop to iterate over the array using funders address and making sure that the every funders funded amount should get reset to 0 again. 
So in First iteration the funder 1 funds get to zero as you said. and it checks the for Loop condition whether the lenght of funders array is less than the current index if not then iterate one more time.

Hence if there are two funders then both funders address to amount funded will get 0 and after that for-loop will end. 
and statements below for loop will start executing.

so for executing ``` funders = new address[](0); ``` this statment the for-loop should end first. In case of two funders it will iterate two times and further statements will get executed.

**Reply by rakesh0x7** [reply]
Hi @DevSwayam , Thanks for the explanation

>In case of two funders it will iterate two times and further statements will get executed.

The thing here is the withdraw function can be called only by the owner. let say owner was funder1 and `funders array` conatins 2 address, one belongs to funder 1 and other belongs to funder 2.

when funder1 calls the withdraw() function after the for-loop execution the funders array is reset i.e funder 2 address will be deleted, which indicates that funder2 can't withdraw his funds cuz his address will be deleted from the `funders array` is this the correct way or only one user can withdraw all the funds in the contract?

hope you got my point!

**Reply by DevSwayam** [reply]
Funder 1 won't be able to call the function if he is not owner and if funder 1 is owner then all funds will be transferred to him.

**Comment by nehal2210** [comment]
hey @rakesh0x7, I think your question is not complete, you did not mention why are you confused, let's say the entire array was reset then what is the problem here?

So I am supposing here that you are confused because, if funders data was reset then how funder can withdraw the amount from the contract?

the answer is, the native currency (ETH) was saved inside the contract. i.e. all the amount submitted by the funders was now associated with the contract address and we use msg. sender to send the amount from contract to msg. sender (function caller).
we are using the funders array only because of tracking.
hope this was helpful.

**Reply by rakesh0x7** [reply]
Thanks for the explanation @nehal2210

**Comment by alymurtazamemon** [comment, marked as answer]
@rakesh0x7 This is the crowdfunding/funding contract where multiple people can give funding to this contract and the authority/owner can get the funds out of the contract whenever they want(no other user has the power to withdraw funds).

Now in the withdrawal, we are resetting both the array and the mapping because we want to remove the previous record once the owner gets the funds out of contract for this round.

Now your question is that it only works for 1 user. So the answer is, it is intentionally created this way so the only owner can get funds out of the contract.

If let's say funders are the owner and any other user, and both gave 1 ETH each and now the contract has 2 ETH deposited. The owner will call the withdraw function which gives the 2 ETH to the owner and reset both array and mappings.

After that new round will be initiated and new funding will be raised.

**Reply by rakesh0x7** [reply]
Thanks for the great explanation @alymurtazamemon 💯 🚀

Now, I understood clearly. Thanks again 😄
